File: rin.Framework/Graphics/Shaders/IShaderManager.cs
namespace rin.Framework.Graphics.Shaders;

public interface IShaderManager : IDisposable
{
    public event Action? OnBeforeDispose;
    public Task<CompiledShader> Compile(IShader shader);
    public IGraphicsShader GraphicsFromPath(string path);
    public IComputeShader ComputeFromPath(string path);
}

File: lib/rin/graphics/IShaderManager.cpp
#include "rin/graphics/IShaderManager.h"
#include <exception>

namespace rin::graphics
{
    void IShaderManager::HandleCompilation(IShader* shader)
    {
        auto context = CreateCompilationContext(shader);
        try
        {
            shader->Compile(context);
            delete context;
        }
        catch (...)
        {
            delete context;
            std::rethrow_exception(std::current_exception());
        }
    }

    io::Task<> IShaderManager::StartCompilation(IShader* shader)
    {
        return _runner.Run<void>([this,shader]()
        {
            return HandleCompilation(shader);
        });
    }
}
